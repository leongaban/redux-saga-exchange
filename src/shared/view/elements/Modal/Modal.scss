@import "~shared/view/styles/variables";
@import "src/shared/view/styles/mixins";

.css-modal {
  position: fixed;
  width: 100%;
  height: 100%;
  bottom: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;

  &__overlay {
    position: fixed;
    z-index: 1000;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: rgba(139, 144, 162, 0.4);
  }

  &__header {
    display: flex;
    padding: 0.79rem 0.79rem 0.79rem 1.5rem;

    /* doiuse-disable */

    cursor: grab;

    /* doiuse-enable */

    text-transform: uppercase;
    font-size: 0.8rem;
    font-weight: 600;
    align-items: center;
    justify-content: space-between;

    @include themed('color', 'headline-1-color');

    &_center {
      justify-content: center;
    }

    &_with-bottom-border {
      border-bottom: 1px solid;

      @include themed('border-color', 'widget-border-color');
    }
  }

  &__cross {
    position: relative;
    width: 0.6rem;
    height: 0.6rem;
    cursor: pointer;
    vertical-align: middle;

    &::after,
    &::before {
      position: absolute;
      left: 0.3rem;
      width: 0.07rem;
      height: 0.6rem;
      content: '';

      @include themed('background-color', 'headline-1-color');
    }

    &::after {
      transform: rotate(45deg);
    }

    &::before {
      transform: rotate(-45deg);
    }

    &-wrapper {
      display: flex;
      width: 1.71rem;
      height: 1.71rem;
      justify-content: center;
      align-items: center;

      &:hover {
        @include themed('background-color', 'menu-entry-hover');
      }
    }
  }

  &__content {
    border: 1px solid;
    box-shadow: 0 0.71rem 0.71rem 0 rgba(53, 64, 82, 0.1);

    @include themed('background-color', 'tooltip-background-color');
    @include themed('border-color', 'widget-border-color');

    &_with-scroll {
      overflow-y: auto;
      max-height: 100vh;
    }
  }

  &__left-part {
    display: flex;
  }

  &__refining-text {
    margin-left: 0.71rem;
    font-weight: bold;

    @include themed('color', 'modal-refining-text-color');
  }
}
