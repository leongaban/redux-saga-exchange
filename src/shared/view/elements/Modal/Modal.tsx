import * as React from 'react';
import block from 'bem-cn';
import ReactModal from 'react-modal';
import Draggable from 'react-draggable';
import { bind } from 'decko';

import './Modal.scss';

interface IProps {
  isOpen: boolean;
  title: React.ReactChild;
  refiningText?: React.ReactNode;
  hasCloseCross?: boolean;
  hasBotttomBorderAtHeader?: boolean;
  withVerticalScroll?: boolean;
  className?: string;
  shouldCloseOnEsc?: boolean;
  onClose?(): void;
}

const b = block('css-modal');

class Modal extends React.PureComponent<IProps> {

  public render() {
    const { isOpen, onClose, withVerticalScroll, className, shouldCloseOnEsc = true } = this.props;

    return (
      <ReactModal
        isOpen={isOpen}
        onRequestClose={onClose}
        className={`${b()} ${className}`}
        overlayClassName={b('overlay')()}
        ariaHideApp={false}
        overlayRef={this.initOverlayRef}
        shouldCloseOnEsc={shouldCloseOnEsc}
      >
        <Draggable
          handle={`.${b('header')()}`}
          bounds={`.${b()}`}
        >
          <div className={b('content', { 'with-scroll': Boolean(withVerticalScroll) })()}>
            {this.renderHeader()}
            {this.props.children}
          </div>
        </Draggable>
      </ReactModal >
    );
  }

  private renderHeader() {
    const { title, hasCloseCross, onClose, hasBotttomBorderAtHeader, refiningText } = this.props;
    const isCenter = !Boolean(hasCloseCross);
    return (
      <div
        className={b('header', {
          'center': isCenter,
          'with-bottom-border': !!hasBotttomBorderAtHeader,
        })()}
      >

        <div className={b('left-part')()}>
          <div className={b('title')()}>
            {title}
          </div>
          {refiningText && (
            <div className={b('refining-text')()}>
              {refiningText}
            </div>
          )}
        </div>
        {hasCloseCross && (
          <div className={b('cross-wrapper')()}>
            <div onClick={onClose} className={b('cross')()} />
          </div>
        )}
      </div>
    );
  }

  @bind
  private initOverlayRef(overlay: HTMLDivElement | null) {
    if (overlay) {
      overlay.addEventListener('touchmove', this.handleOverlayTouchMove);
    }
  }

  @bind
  private handleOverlayTouchMove(e: TouchEvent) {
    // disable scrolling
    e.preventDefault();
  }
}

export { IProps as IModalProps };
export default Modal;
