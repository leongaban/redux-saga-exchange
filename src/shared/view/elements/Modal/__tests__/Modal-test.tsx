import * as React from 'react';
import block from 'bem-cn';
import { shallow } from 'enzyme';
import Modal from '../Modal';

describe('(Shared) component', () => {

    describe('Modal', () => {

        const b = block('css-modal');

        it('should render', () => {
            const component = shallow(
            <Modal title={'header'} isOpen={true} onClose={jest.fn()} />,
          );

            expect(component.debug()).toMatchSnapshot();
        });

        it('hasCloseCross: true, must contain close cross ', () => {
            const component = shallow(
            <Modal title={'header'} isOpen={true} onClose={jest.fn()} hasCloseCross={true} />,
          );

            expect(component.debug()).toMatchSnapshot();
        });

        it('click on cross must call onClose', () => {
            const onClose = jest.fn();
            const component = shallow(
                <Modal title={'header'} isOpen={true} onClose={onClose} hasCloseCross={true} />,
            );

            component.find(`.${b('cross')}`).simulate('click');
            expect(onClose).toBeCalled();
        });
    });

});
