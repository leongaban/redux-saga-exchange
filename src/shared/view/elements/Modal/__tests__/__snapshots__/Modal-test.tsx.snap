// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Shared) component Modal hasCloseCross: true, must contain close cross  1`] = `
"<Modal isOpen={true} onRequestClose={[Function: mockConstructor]} className=\\"css-modal\\" overlayClassName=\\"css-modal__overlay\\" ariaHideApp={false} portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} parentSelector={[Function: parentSelector]}>
  <Draggable handle=\\".css-modal__header\\" bounds=\\".css-modal\\" allowAnyClick={false} cancel={{...}} disabled={false} enableUserSelectHack={true} offsetParent={{...}} grid={{...}} transform={{...}} onStart={[Function: onStart]} onDrag={[Function: onDrag]} onStop={[Function: onStop]} onMouseDown={[Function: onMouseDown]} axis=\\"both\\" defaultClassName=\\"react-draggable\\" defaultClassNameDragging=\\"react-draggable-dragging\\" defaultClassNameDragged=\\"react-draggable-dragged\\" defaultPosition={{...}} position={{...}}>
    <div className=\\"css-modal__content\\">
      <div className=\\"css-modal__header\\">
        <div className=\\"css-modal__left-part\\">
          <div className=\\"css-modal__title\\">
            header
          </div>
        </div>
        <div className=\\"css-modal__cross-wrapper\\">
          <div onClick={[Function: mockConstructor]} className=\\"css-modal__cross\\" />
        </div>
      </div>
    </div>
  </Draggable>
</Modal>"
`;

exports[`(Shared) component Modal should render 1`] = `
"<Modal isOpen={true} onRequestClose={[Function: mockConstructor]} className=\\"css-modal\\" overlayClassName=\\"css-modal__overlay\\" ariaHideApp={false} portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} parentSelector={[Function: parentSelector]}>
  <Draggable handle=\\".css-modal__header\\" bounds=\\".css-modal\\" allowAnyClick={false} cancel={{...}} disabled={false} enableUserSelectHack={true} offsetParent={{...}} grid={{...}} transform={{...}} onStart={[Function: onStart]} onDrag={[Function: onDrag]} onStop={[Function: onStop]} onMouseDown={[Function: onMouseDown]} axis=\\"both\\" defaultClassName=\\"react-draggable\\" defaultClassNameDragging=\\"react-draggable-dragging\\" defaultClassNameDragged=\\"react-draggable-dragged\\" defaultPosition={{...}} position={{...}}>
    <div className=\\"css-modal__content\\">
      <div className=\\"css-modal__header css-modal__header_center\\">
        <div className=\\"css-modal__left-part\\">
          <div className=\\"css-modal__title\\">
            header
          </div>
        </div>
      </div>
    </div>
  </Draggable>
</Modal>"
`;
